### PR TITLE
New hash data format

### PR DESCRIPTION
**Pull request by Vojtech-Sassmann**
* We needed to adjust to generated format so it can be easily proccessed
in scripts.
* The generator logic should be refactored since the generated format
has changed and it could be much better designed.

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java [hunk 1]
```diff
@@ -380,30 +380,24 @@
...
-	 *          }
-	 *        ]
-	 *      },
-	 *      {
-	 *        ** resource2 **
-	 *        ...
-	 *      }
-	 *    ]
+	 *   "1": {    ** facility id **
+	 *     members: {    ** all members on resource **
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java:384]: These are all members on the facility, right?

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java [hunk 2]
```diff
@@ -380,30 +380,24 @@
...
+	 *        "4" : 5,    ** member id : user id **
+	 *        "6" : 7,    ** member id : user id **
+	 *       ...
+	 *     }
+	 *     children: [
+	 *       "2": {    ** resource id **
+	 *         children: []
+	 *         members: {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java:392]: These should be all members on the resource 2.

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java (hunk 1 above)
Comment (stavamichal) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java:384]: The same for all other places where this Javadoc is used.

**Review by stavamichal — CHANGES_REQUESTED** [review]

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java (hunk 1 above)
Comment (Vojtech-Sassmann) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java]: Yes, I have updated the javadocs.

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java (hunk 2 above)
Comment (Vojtech-Sassmann) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/ServicesManager.java]: I have added a comment in the javadocs.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java [hunk 3]
```diff
@@ -54,6 +56,7 @@
 	private final Service service;
 	private final Facility facility;
 	private final GenDataProvider dataProvider;
+	private final Set<Member> allMembers = new HashSet<>();
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java:59]: Shouldn't ``hashes: [...hashes...]`` be removed also from this class javadoc? It was removed from services manager methods API docs.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java [hunk 4]
```diff
@@ -68,6 +68,7 @@
 	private final Facility facility;
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java:68]: Shouldn't ``hashes: [...hashes...]`` be removed also from this class javadoc? It was removed from services manager methods API docs.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java [hunk 5]
```diff
@@ -94,18 +99,19 @@ private GenDataNode getDataForResource(Resource resource) {
 		} else {
 			members = sess.getPerunBl().getResourcesManagerBl().getAllowedMembers(sess, resource);
 		}
+		allMembers.addAll(members);
 
 		dataProvider.loadResourceAttributes(resource, members, false);
 
 		List<String> resourceAttrHashes = dataProvider.getResourceAttributesHashes(resource, false);
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java:106]: Variable ``resourceAttrHashes`` is not used.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java [hunk 6]
```diff
@@ -72,19 +75,21 @@ public HashedGenData generateData() {
 		List<Resource> resources =
 				sess.getPerunBl().getFacilitiesManagerBl().getAssignedResources(sess, facility, null, service);
 
-		List<GenDataNode> childNodes = resources.stream()
-				.map(this::getDataForResource)
-				.collect(Collectors.toList());
+		Map<Integer, GenDataNode> childNodes = resources.stream()
+				.collect(toMap(Resource::getId, this::getDataForResource));
 
 		List<String> facilityAttrHashes = dataProvider.getFacilityAttributesHashes();
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java:81]: ``facilityAttrHashes`` is not used anywhere.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java [hunk 7]
```diff
@@ -106,26 +109,27 @@ private GenDataNode getDataForResource(Resource resource) {
 		} else {
 			members = sess.getPerunBl().getResourcesManagerBl().getAllowedMembers(sess, resource);
 		}
+		allMembers.addAll(members);
 
 		dataProvider.loadResourceAttributes(resource, members, true);
 
 		List<String> resourceAttrHashes = dataProvider.getResourceAttributesHashes(resource, true);
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java:116]: ``resourceAttrHashes`` is not used.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java [hunk 8]
```diff
@@ -84,19 +85,21 @@ public HashedGenData generateData() {
 
 		List<Resource> resources = sess.getPerunBl().getFacilitiesManagerBl().getAssignedResources(sess, facility, null, service);
 
-		List<GenDataNode> childNodes = resources.stream()
-				.map(this::getDataForResource)
-				.collect(Collectors.toList());
+		Map<Integer, GenDataNode> childNodes = resources.stream()
+				.collect(toMap(Resource::getId, this::getDataForResource));
 
 		List<String> facilityAttrHashes = dataProvider.getFacilityAttributesHashes();
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java:91]: ``facilityAttrHashes`` is not used.

**Review by zlamalp — CHANGES_REQUESTED** [review]

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java (hunk 3 above)
Comment (Vojtech-Sassmann) [reply perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java:53]: Yes, I have removed it.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java (hunk 4 above)
Comment (Vojtech-Sassmann) [reply perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java:60]: Yes, I have removed it.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java (hunk 5 above)
Comment (Vojtech-Sassmann) [reply perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java]: Variable removed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java (hunk 6 above)
Comment (Vojtech-Sassmann) [reply perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/HierarchicalHashedDataGenerator.java]: Variable removed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java (hunk 7 above)
Comment (Vojtech-Sassmann) [reply perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java]: Variable removed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java (hunk 8 above)
Comment (Vojtech-Sassmann) [reply perun-core/src/main/java/cz/metacentrum/perun/core/provisioning/GroupsHashedDataGenerator.java]: Variable removed.

**Review by zlamalp — APPROVED** [review]

**Review by stavamichal — APPROVED** [review]

**Pull request merged**